networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/discovery/ConnectionManager.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.networking.p2p.discovery;
+
+import java.time.Duration;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import tech.pegasys.artemis.networking.p2p.network.P2PNetwork;
+import tech.pegasys.artemis.networking.p2p.peer.Peer;
+import tech.pegasys.artemis.service.serviceutils.Service;
+import tech.pegasys.artemis.util.async.AsyncRunner;
+import tech.pegasys.artemis.util.async.SafeFuture;
+
+public class ConnectionManager extends Service {
+  private static final Logger LOG = LogManager.getLogger();
+  private static final Duration RECONNECT_TIMEOUT = Duration.ofSeconds(20);
+  private static final Duration DISCOVERY_INTERVAL = Duration.ofSeconds(30);
+  private final AsyncRunner asyncRunner;
+  private final P2PNetwork<? extends Peer> network;
+  private final List<String> staticPeers;
+  private final DiscoveryService discoveryService;
+
+  public ConnectionManager(
+      final DiscoveryService discoveryService,
+      final AsyncRunner asyncRunner,
+      final P2PNetwork<? extends Peer> network,
+      final List<String> staticPeers) {
+    this.asyncRunner = asyncRunner;
+    this.network = network;
+    this.staticPeers = staticPeers;
+    this.discoveryService = discoveryService;
+  }
+
+  @Override
+  protected SafeFuture<?> doStart() {
+    staticPeers.forEach(this::createPersistentConnection);
+    connectToKnownPeers();
+    searchForPeers().reportExceptions();
+    return SafeFuture.COMPLETE;
+  }
+
+  private void connectToKnownPeers() {
+    discoveryService.streamKnownPeers().forEach(this::attemptConnection);
```

Comment:
Logged https://pegasys1.atlassian.net/browse/BC-284 and https://pegasys1.atlassian.net/browse/BC-285 to limit the total connected peers and avoid connecting to duplicates.

As discussed elsewhere, we probably want to have a target number of peers with some variance allowed above and below that before we take action. That makes a lot of things easier to implement but particularly allows us to accept connections from new peers then evaluate if they are better than existing ones to decide who to drop instead of always rejecting the new peer.